src/Test/L0/Util/RepositoryUtilL0.cs

```diff
@@ -0,0 +1,167 @@
+        [Trait("Level", "L0")]
+        [Trait("Category", "Common")]
+        public void GetCloneDirectory_STRING_should_throw_on_null()
+        {
+            using (TestHostContext hc = new TestHostContext(this))
+            {
+                Tracing trace = hc.GetTrace();
+
+                Assert.Throws<ArgumentNullException>(() => RepositoryUtil.GetCloneDirectory((string)null));
+            }
+        }
+
+        [Fact]
+        [Trait("Level", "L0")]
+        [Trait("Category", "Common")]
+        public void GetCloneDirectory_STRING_should_return_proper_value_when_called()
+        {
+            // These test cases were inspired by the test cases that git.exe uses
+            // see https://github.com/git/git/blob/53f9a3e157dbbc901a02ac2c73346d375e24978c/t/t5603-clone-dirname.sh#L21
+
+            using (TestHostContext hc = new TestHostContext(this))
+            {
+                Tracing trace = hc.GetTrace();
+
+                // basic syntax with bare and non-bare variants
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("host:foo"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("host:foo.git"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("host:foo/.git"));
+
+                // similar, but using ssh URL rather than host:path syntax
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("ssh://host/foo"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("ssh://host/foo.git"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("ssh://host/foo/.git"));
+
+                // we should remove trailing slashes and .git suffixes
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("ssh://host/foo/"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("ssh://host/foo///"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("ssh://host/foo/.git/"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("ssh://host/foo.git/"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("ssh://host/foo.git///"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("ssh://host/foo///.git/"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("ssh://host/foo/.git///"));
+
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("host:foo/"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("host:foo///"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("host:foo.git/"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("host:foo/.git/"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("host:foo.git///"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("host:foo///.git/"));
+                Assert.Equal("foo", RepositoryUtil.GetCloneDirectory("host:foo/.git///"));
+
```

Comment:
should we add tests for "host:bar/foo"? Is that a possibility?

Reply:
I can certainly add it. Anything is possible. I took my test cases directly from git.exe. I didn't add any others, even though there are lots more cases.